[5.1] Added 'postJson', 'putJson', 'patchJson', 'deleteJson' methods to CrawlerTrait

samlev opened the pull request:
A method to allow posting 'application/json' content to a URI. This
mimics how SPAs built on AngularJS send data to a URI.

A recent Laravel update made tests fail when data was sent as POST
parameters instead of POST body with the content type set to
'application/json', because the Request would no longer register
the POST parameters as variables (instead looking for POST body JSON)

This update allows tests to easily post JSON content in a format that
is expected by the Request.

Commit:
Added '*Json' methods to CrawlerTrait.php

A method to allow sending 'application/json' content to a URI. This
mimics how SPAs built on AngularJS send data to a URI.

A recent Laravel update made tests fail when data was sent as POST
parameters instead of POST body with the content type set to
'application/json', because the Request would no longer register
the POST parameters as variables (instead looking for POST body JSON)

This update allows tests to easily fire GET, POST, PUT, FETCH or DELETE
requests with JSON content in a format that is expected by the Request.

GrahamCampbell commented on src/Illuminate/Foundation/Testing/CrawlerTrait.php:
```diff
@@ -93,6 +93,26 @@ public function post($uri, array $data = [], array $headers = [])
     /**
+     * Visit the given URI with a POST request with content type of application/json.
+     *
+     * @param string $uri
```
please fix cs

samlev replied:
Sorry - what needs to be done? Got any links? I was just following the style in the rest of the file.

GrahamCampbell replied:
> I was just following the style in the rest of the file.

Look closer at the positioning of spaces, and you'll see you've not matched the rest of the file.

samlev replied:
Ahh ok, gotcha.

GrahamCampbell commented on src/Illuminate/Foundation/Testing/CrawlerTrait.php:
```diff
@@ -93,6 +93,26 @@ public function post($uri, array $data = [], array $headers = [])
+     * @param array $headers
+     * @return $this
+     */
+    public function postJson($uri, $data = [], $headers = [])
```
please add the missing typehints

GrahamCampbell commented:
Just giving the CI services a kick. :)

kidshenlong commented:
Thanks for this! Quick question though, does this need to be a new function? Couldn't we overload the current method or maybe add an additional parameter? 
```
public function post($uri, array $data = [], array $headers = [], $isJson = false)
```
Also should this pull request be extended to include PUT and PATCH?

samlev commented:
An 'isJson' flag might be appropriate, but I'm not too keen to change the interface of existing functions. PUT and PATCH should probably be included, too.

I'll update and re-push.

GrahamCampbell commented:
Not sure I like this. :-1:

samlev commented:
Happy to change it to something that you feel is more appropriate.

GrahamCampbell commented:
It was the whole idea, not the implementation. Ping @taylorotwell.

crynobone commented on src/Illuminate/Foundation/Testing/CrawlerTrait.php:
```diff
@@ -93,6 +93,26 @@ public function post($uri, array $data = [], array $headers = [])
+    public function postJson($uri, array $data = [], array $headers = [])
+    {
+        $headers['CONTENT_TYPE'] = 'application/json';
+        $headers['Accept'] = 'application/json';
```
While it is rare, wouldn't it be possible to post json (`Content-Type=application/json`) and at the same time `Accept=application/xml` etc.

samlev replied:
Yeah, that's a fair point. The 'accept' param should only be set if it's not otherwise set.

kidshenlong replied:
I was thinking this too. Wouldn't it be better to be a bit more general and let the user still specify accept headers, etc for these cases?

samlev commented:
From what I can tell, there was a change to the Symphony 'Request' which is imperceptible when actually firing requests, but when running tests it causes the Request class to fail to correctly catch the passed data (in short, my unit tests broke overnight after a composer update).

My best guess for *why* it's breaking is because the Request object sees 'CONTENT_TYPE = application/json' and tries to read the content body instead of the post params. This is (imo) important to test because that's how angularjs actually sends data.

The other issue that I ran into while testing is that this update doesn't actually solve the problem entirely either. If you have nested data (say, an array in the POST), then this method only seems to collect top-level vars. e.g.:

```php
$data = [
    'var1' => 'foo',
    'var2 => [
        'bar' => 'baz',
        'fizz' => 'buzz'
    ]
]
```

Posting with body content would make 'var1' exist, but 'var2' would be empty. Posting with parameters would make 'var2' data exist, but 'var1' wouldn't. To get it to recognise both top level *and* nested vars, I needed to post it to both params and content:

```php
$this->call('POST', $uri, $data, [], [], $server, json_encode($data));
```

That's not ideal, so I'm digging into the Symphony source to see what it's actually expecting from a real/live post. Current working theory is that it's actually expecting a reource object (but that doesn't seem to be working, either)

samlev commented:
Opened issue #9970 to explain what I'm actually trying to fix here.

samlev commented:
Added related pull request #9971 that fixes another related bug.

samlev commented:
I've updated this to include CONTENT_LENGTH, and I've added `getJson` for completeness (as it's technically possible to [send a payload with a GET request](http://tools.ietf.org/html/rfc7231#section-4.3.1), although uncommon).

GrahamCampbell commented on src/Illuminate/Foundation/Testing/CrawlerTrait.php:
```diff
@@ -76,6 +76,28 @@ public function get($uri, array $headers = [])
+    {
+        $content = json_encode($data);
+        $headers['CONTENT_TYPE'] = 'application/json';
+        $headers['CONTENT_LENGTH'] = strlen($content);
```
you should probably use the multibyte function here

GrahamCampbell commented on src/Illuminate/Foundation/Testing/CrawlerTrait.php:
```diff
@@ -76,6 +76,28 @@ public function get($uri, array $headers = [])
+        $content = json_encode($data);
+        $headers['CONTENT_TYPE'] = 'application/json';
+        $headers['CONTENT_LENGTH'] = strlen($content);
+        $headers['Accept'] = $headers['Accept'] ?: 'application/json';
```
you shouldn't reasign this variable if you don't need to

GrahamCampbell replied:
better to have an if statement here

GrahamCampbell replied:
infact, you need to to avoid a fatal error...

GrahamCampbell replied:
http://3v4l.org/K242V

samlev replied:
It's not a fatal error (at least not in any version of PHP that Laravel supports), but it is potentially an "undefined index" notice.

I'll update that, but I feel that an expected 'Accept' header for json calls is appropriate (and easy to overwrite).

samlev commented:
Squashed to 3 commits

samlev commented:
Re-thought, and squashed to 1 commit.

GrahamCampbell commented on src/Illuminate/Foundation/Testing/CrawlerTrait.php:
```diff
@@ -76,6 +76,30 @@ public function get($uri, array $headers = [])
+        $content = json_encode($data);
+        $headers['CONTENT_TYPE'] = 'application/json';
+        $headers['CONTENT_LENGTH'] = mb_strlen($content);
+        if (isset($headers['Accept']) === false) {
```
don't need `=== false`

GrahamCampbell replied:
`(! isset())`

samlev replied:
Yeah, I did that originally, then switched it to === false because that seemed to fit better with the rest of the style of the code. I can switch it back.

GrahamCampbell replied:
the rest of the code style definitely is`(! isset())`.

lukasgeiter replied:
@samlev Typesafe comparison (`===`) is used when we can't totally rely on the type of the value that is checked to avoid false positives. However, `isset()` will always return a boolean so we don't need that.

GrahamCampbell commented on src/Illuminate/Foundation/Testing/CrawlerTrait.php:
```diff
@@ -76,6 +76,30 @@ public function get($uri, array $headers = [])
+    {
+        $content = json_encode($data);
+        $headers['CONTENT_TYPE'] = 'application/json';
+        $headers['CONTENT_LENGTH'] = mb_strlen($content);
```
leave an extra blank line after this

GrahamCampbell replied:
also, the mb_strlen thing still looks wrong

GrahamCampbell replied:
you want the actual number of bytes, not the number of characters

samlev replied:
That's unfortunately all we have to work with. The number of bytes to expect *is* the length of the string, and I'm unaware of any way in PHP to find the number of bytes in a string other than getting it's length.

GrahamCampbell replied:
> That's unfortunately all we have to work with. 

No it's not. `mb_strlen($string, '8bit')` will give you the number of bytes.

samlev commented:
Pushed a new version with all the style updates

kidshenlong commented:
Any news on this?

xiphiaz commented:
This is needed to be able to test `$request->json()` calls in the controller. You can't just encode it yourself, as the crawler methods are typecasted to array. Currently the only way make it work is to overwrite the trait methods.

xiphiaz commented:
@samlev You should set the header casing to follow http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.13

e.g. `Content-Length` not `CONTENT_LENGTH`

samlev commented:
Internally, it uses `CONTENT_LENGTH` because in a real request, it's fetching the details from `$_SERVER` (where it's reported as either `CONTENT_LENGTH` or `HTTP_CONTENT_LENGTH`)

Because we're faking calls, it's not actually creating a usable `$_SERVER` array, so if we set `Content-Length` it would actually get missed by Symfony's `ServerBag`/`Request` objects (which the Laravel `Request` is built off.

See: [`Symfony\Component\HttpFoundation\ServerBag->getHeaders()`](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/HttpFoundation/ServerBag.php#L31) and [`Symfony\Component\HttpFoundation\Request->createFromGlobals()`](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/HttpFoundation/Request.php#L292)

xiphiaz commented:
@samlev you'll see in the `$this->transformHeadersToServerVars` that it actually translates header keys to `HTTP_*` vars, so it is better to use the actual intended http header for consistency with the input param.

xiphiaz commented:
@samlev I've patched my code with your PR, as I can't wait for this to be merged (if it ever does). FYI I extracted the duplicate code with the following function:
```
/**
     * @param array $headers
     * @param $content
     * @return array
     */
    protected function addJsonHeaders(array $headers, $content)
    {
        $headers['Content-Type'] = 'application/json';
        $headers['Content-Length'] = mb_strlen($content, '8bit');
        if (!isset($headers['Accept'])) {
            $headers['Accept'] = 'application/json';
            return $headers;
        }
        return $headers;
    }
```

patchJson would then become
```
public function patchJson($uri, array $data = [], array $headers = [])
    {
        $content = json_encode($data);

        $headers = $this->addJsonHeaders($headers, $content);
        $server = $this->transformHeadersToServerVars($headers);
        $this->call('PATCH', $uri, [], [], [], $server, $content);
        return $this;
    }
```

samlev commented:
@xiphiaz I suppose it's academic at this point, as the headers get transformed before they ever get sent to `call`. It also appears to open up a situation where whichever value is set last for `Content-Length`, `CONTENT_LENGTH` or any other variation of it would become authoritative. Not that that's really a bug to discuss here, and that's probably really more "user error" than anything else (as it is, we would be adding the last `CONTENT_LENGTH`, so it would probably be gospel anyway)

samlev commented:
@xiphiaz Also, it's good that we can just add these functions to TestCase.php instead of having to overwrite CrawlerTrait in the vendors file (or having to run from a clone). Once these updates get rolled in (*if* these updates ever get rolled in), I'll just be deleting my copy from TestCase.php and going from there.

xiphiaz commented:
@samlev yep I'm taking that approach now :) I'm not sure why @GrahamCampbell is resistance to this PR, it make it impossible to test the `Request` function `$request->json()`.

kidshenlong commented:
Any news on the status of this @GrahamCampbell? Thanks!

GrahamCampbell commented:
Not up to me. Up to @taylorotwell.

samlev commented:
Ping @taylorotwell - what can I do to get this through?

taylorotwell commented:
I'm curious why you are using the function $request->json() and not $request->get()?

samlev commented:
In what context?

This PR is about being able to send json body content to a view (which is how libraries like angular send content).

In the controllers themselves, using `$request->get()` *won't work* in that instance, because `get` comes from the symphony request model, which doesn't know about/support json as body content.

taylorotwell commented:
Ah, sorry. I meant `$request->input()`. I use Laravel with Angular / Vue all the time and just always use $request->input() and it works fine?

samlev commented:
Yeah, it does (there was another PR to make `$request->get()` work with json, too). This also adds some extra headers which simulate an "expects json" style response (again, similar to angular).

taylorotwell commented:
I guess I don't understand why we don't just have the $request->json() method call $request->input() internally. The two methods should behave indentically.

samlev commented:
I think (I could be wrong, it's been a couple of weeks since I looked at it), it works the exact opposite to that - `$request->input()` calls `$request->json()` if `$request->isJson()`.

taylorotwell commented:
Ah OK, I see that now.

taylorotwell commented:
I should have removed $request->json() as a public method so people just use `input`, then we wouldn't have this problem :)

taylorotwell commented:
You really should just use `$request->input()` all the time so you don't have to worry about it.

samlev commented:
The issue that I was having when I ran into this initially was that I was using `$request->get()` (which, as I mentioned, is unaware of json) instead of `$request->input()`.

The other side that I mentioned is that the response may be different for json requests than it would be for straight get/post/whatever requests (i.e. you may want to respond with json for a request that expects json, and a blade template for other requests).

One important part where that's noticeable is error responses (e.g. authentication errors). For a json request, you would expect the response to be returned as a json object that you can consume on the client, but by default it returns an error page which is effectively useless to an API call.

taylorotwell commented:
I added a simple `json` method to the CrawlerTrait.

xiphiaz commented:
Code link for future reference https://github.com/laravel/framework/commit/d8e351e17bc25fa7c86bca0e250fde889167f3bb

Note that you have to pass the method eg `$this->json('GET', '/path/to/endpoint/', ['foo' => 'bar']);`
